<launch>

    <param name="KeyFramePath" type="string" value="/home/i/VLIO/src/VLIO/PGO/keyframes"/>

    <param name="scan_line" type="int" value="16" />

    <!-- if 1, do mapping 10 Hz, if 2, do mapping 5 Hz. Suggest to use 1, it will adjust frequence automaticlly -->
    <param name="mapping_skip_frame" type="int" value="1" />

    <!-- remove too closed points -->
    <param name="minimum_range" type="double" value="0.1"/>

    <param name="mapping_line_resolution" type="double" value="0.2"/>
    <param name="mapping_plane_resolution" type="double" value="0.4"/>

    <param name="mapviz_filter_size" type="double" value="0.4"/>

    <!-- SC-A-LOAM -->
    <param name="keyframe_meter_gap" type="double" value="5"/> 
    <param name="keyframe_deg_gap" type="double" value="10.0"/> 

    <!-- Scan Context -->
    <param name="sc_dist_thres" type="double" value="0.4"/> <!-- SC-A-LOAM, if want no outliers, use 0.1-0.15 -->
    <param name="sc_max_radius" type="double" value="40"/> <!-- 20 or 40 for indoor -->
    <!-- <param name="sc_max_radius" type="double" value="80.0"/> for outdoor -->

    <!-- loop closure -->
    <param name="historyKeyframeSearchRadius" value="1.5"/>
    <param name="historyKeyframeSearchTimeDiff" value="30.0"/>
    <param name="historyKeyframeSearchNum" value="20"/>
    <param name="speedFactor" value="1"/>           <!-- 倍速因子 -->
    <param name="loopClosureFrequency" value="4.0"/> <!-- 应该相对于输入数据的频率 成比例 -->
    <param name="graphUpdateFrequency" value="2.0"/>
    <param name="graphUpdateTimes" value="5"/> <!-- 每次更新，ism->update的测试 -->
    <param name="loopNoiseScore" value="0.1"/> <!-- 0.02 -->
    <param name="vizmapFrequency" value="10"/>
    <param name="loopFitnessScoreThreshold" value="0.3"/> <!-- ICP score -->
    
    
    <param name="lidar_type" type="string" value="VLP16"/>
    <remap from="/velodyne_points" to="/velodyne_points"/>
    
    <!-- input from FASTLIO2 -->
    <remap from="/aft_mapped_to_init" to="/Odometry"/>
    <remap from="/velodyne_cloud_registered_local" to="/cloud_registered_body"/>
    <!-- <remap from="/cloud_for_scancontext" to="/cloud_registered_lidar"/>   because ScanContext requires lidar-ego-centric coordinate for the better performance -->

    <!-- utils -->
    <param name="save_directory" type="string" value="/home/i/VLIO/src/VLIO/PGO/data/"/>  <!-- CHANGE THIS and end with / -->

    <!-- nodes -->
    <node pkg="aloam_velodyne" type="alaserPGO" name="alaserPGO" output="screen" /> <!-- Scan Context-based PGO -->

    <!-- visulaization -->
    <arg name="rviz" default="true" />
    <group if="$(arg rviz)">
        <node launch-prefix="nice" pkg="rviz" type="rviz" name="rviz_sc_pgo" args="-d $(find aloam_velodyne)/rviz_cfg/aloam_velodyne.rviz" />
    </group>

</launch>
